Querier panic when marshaling frontendv2pb.(*QueryResultRequest)

**Describe the bug**
The querier get panics when marshaling `frontendv2pb.(*QueryResultRequest)` in my test cluster.

**To Reproduce**
I reproduce it in my local [branch test code](https://github.com/SungJin1212/cortex/blob/Fix-query-scheduler-bug/pkg/querier/worker/scheduler_processor_test.go)

The panic scenario I guess is:
1. scheduler processor call querier [handler](https://github.com/cortexproject/cortex/blob/release-1.18/pkg/querier/worker/scheduler_processor.go#L179)
2. querier handler create go-routines to handle queries. In go-routines, it increments [query-stats](https://github.com/cortexproject/cortex/blob/release-1.18/pkg/querier/blocks_store_queryable.go#L755-L761).
3. scheduler processor call [QueryResult](https://github.com/cortexproject/cortex/blob/release-1.18/pkg/querier/worker/scheduler_processor.go#L212) 
4. the panic could happen when the query-stats values are incremented in `fetchSeriesFromStores` (by not yet exited go routines) while progressing `(*QueryResultRequest).MarshalToSizedBuffer` because the same query-stat objects are used.

**Expected behavior**
Not panic

**Environment:**
 - Infrastructure: kubernetes

**Additional Context**
<!--  Additional relevant info which can help us debug this issue easily like Logs, Configuration etc. -->
logs:
```
panic: runtime error: index out of range [-3]

goroutine 7055 [running]:
github.com/cortexproject/cortex/pkg/frontend/v2/frontendv2pb.encodeVarintFrontend(...)
	/go/src/github.com/cortexproject/cortex/pkg/frontend/v2/frontendv2pb/frontend.pb.go:402
github.com/cortexproject/cortex/pkg/frontend/v2/frontendv2pb.(*QueryResultRequest).MarshalToSizedBuffer(0xc0027eac90, {0xc002be8000, 0xa5, 0xa5?})
	/go/src/github.com/cortexproject/cortex/pkg/frontend/v2/frontendv2pb/frontend.pb.go:368 +0x2c7
github.com/cortexproject/cortex/pkg/frontend/v2/frontendv2pb.(*QueryResultRequest).Marshal(0xc0027eac90)
	/go/src/github.com/cortexproject/cortex/pkg/frontend/v2/frontendv2pb/frontend.pb.go:326 +0x4d
google.golang.org/protobuf/internal/impl.legacyMarshal({{}, {0x341faf0, 0xc002bd2250}, {0x0, 0x0, 0x0}, 0x0})
	/go/src/github.com/cortexproject/cortex/vendor/google.golang.org/protobuf/internal/impl/legacy_message.go:411 +0xf2
google.golang.org/protobuf/proto.MarshalOptions.marshal({{}, 0x90?, 0xac?, 0x7e?}, {0x0, 0x0, 0x0}, {0x341faf0, 0xc002bd2250})
	/go/src/github.com/cortexproject/cortex/vendor/google.golang.org/protobuf/proto/encode.go:195 +0x35d
google.golang.org/protobuf/proto.Marshal({0x33db240, 0xc002bd2250})
	/go/src/github.com/cortexproject/cortex/vendor/google.golang.org/protobuf/proto/encode.go:111 +0x50
google.golang.org/grpc/encoding/proto.codec.Marshal({}, {0x2cbe5e0?, 0xc0027eac90?})
	/go/src/github.com/cortexproject/cortex/vendor/google.golang.org/grpc/encoding/proto/proto.go:47 +0x52
google.golang.org/grpc.encode({0x7fd8963b7528?, 0x4c9f480?}, {0x2cbe5e0?, 0xc0027eac90?})
	/go/src/github.com/cortexproject/cortex/vendor/google.golang.org/grpc/rpc_util.go:647 +0x3e
google.golang.org/grpc.prepareMsg({0x2cbe5e0?, 0xc0027eac90?}, {0x7fd8963b7528?, 0x4c9f480?}, {0x0, 0x0}, {0x0, 0x0})
	/go/src/github.com/cortexproject/cortex/vendor/google.golang.org/grpc/stream.go:1772 +0xc7
google.golang.org/grpc.(*clientStream).SendMsg(0xc0020ceea0, {0x2cbe5e0, 0xc0027eac90})
	/go/src/github.com/cortexproject/cortex/vendor/google.golang.org/grpc/stream.go:894 +0xd7
google.golang.org/grpc.invoke({0x33f9868?, 0xc002baacc0?}, {0x2e3dc84?, 0x0?}, {0x2cbe5e0, 0xc0027eac90}, {0x2c917c0, 0x4c9f480}, 0x2b8b940?, {0xc002ef85e0, ...})
	/go/src/github.com/cortexproject/cortex/vendor/google.golang.org/grpc/call.go:70 +0x9f
github.com/cortexproject/cortex/pkg/querier/worker.(*schedulerProcessor).createFrontendClient.PrometheusGRPCUnaryInstrumentation.func1({0x33f9868, 0xc002baacc0}, {0x2e3dc84, 0x2c}, {0x2cbe5e0, 0xc0027eac90}, {0x2c917c0, 0x4c9f480}, 0xc00261cc08, 0x2fa02d0, ...)
	/go/src/github.com/cortexproject/cortex/pkg/util/middleware/grpc.go:20 +0x10b
github.com/cortexproject/cortex/pkg/util/grpcclient.(*Config).DialOption.ChainUnaryClient.func6.1({0x33f9868?, 0xc002baacc0?}, {0x2e3dc84?, 0x0?}, {0x2cbe5e0?, 0xc0027eac90?}, {0x2c917c0?, 0x4c9f480?}, 0xc002baac00?, {0xc002ef85e0, ...})
	/go/src/github.com/cortexproject/cortex/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:113 +0x91
github.com/weaveworks/common/middleware.ClientUserHeaderInterceptor({0x33f9868?, 0xc002baac30?}, {0x2e3dc84, 0x2c}, {0x2cbe5e0, 0xc0027eac90}, {0x2c917c0, 0x4c9f480}, 0xc00261cc08, 0xc002baab40, ...)
	/go/src/github.com/cortexproject/cortex/vendor/github.com/weaveworks/common/middleware/grpc_auth.go:17 +0xd7
github.com/cortexproject/cortex/pkg/util/grpcclient.(*Config).DialOption.ChainUnaryClient.func6.1({0x33f9868?, 0xc002baac30?}, {0x2e3dc84?, 0xc002babf80?}, {0x2cbe5e0?, 0xc0027eac90?}, {0x2c917c0?, 0x4c9f480?}, 0xc000092030?, {0xc002ef85e0, ...})
	/go/src/github.com/cortexproject/cortex/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:113 +0x91
github.com/opentracing-contrib/go-grpc.OpenTracingClientInterceptor.func1({0x33f9868, 0xc002baa060}, {0x2e3dc84, 0x2c}, {0x2cbe5e0, 0xc0027eac90}, {0x2c917c0, 0x4c9f480}, 0xc00261cc08, 0xc002baab70, ...)
	/go/src/github.com/cortexproject/cortex/vendor/github.com/opentracing-contrib/go-grpc/client.go:62 +0x44e
github.com/cortexproject/cortex/pkg/util/grpcclient.(*Config).DialOption.ChainUnaryClient.func6({0x33f9868, 0xc002baa060}, {0x2e3dc84, 0x2c}, {0x2cbe5e0, 0xc0027eac90}, {0x2c917c0, 0x4c9f480}, 0xc00261cc08, 0x2fa02d0, ...)
	/go/src/github.com/cortexproject/cortex/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:116 +0x1d3
google.golang.org/grpc.(*ClientConn).Invoke(0xc00261cc08, {0x33f9868?, 0xc002baa060?}, {0x2e3dc84?, 0x7fd8dd280f18?}, {0x2cbe5e0?, 0xc0027eac90?}, {0x2c917c0?, 0x4c9f480?}, {0x0, ...})
	/go/src/github.com/cortexproject/cortex/vendor/google.golang.org/grpc/call.go:35 +0x205
github.com/cortexproject/cortex/pkg/frontend/v2/frontendv2pb.(*frontendForQuerierClient).QueryResult(0x40e81f?, {0x33f9868?, 0xc002baa060?}, 0x410b00?, {0x0?, 0x2a571a0?, 0x1?})
	/go/src/github.com/cortexproject/cortex/pkg/frontend/v2/frontendv2pb/frontend.pb.go:268 +0x5d
github.com/cortexproject/cortex/pkg/querier/worker.(*schedulerProcessor).runRequest(0xc001072c60, {0x33f9868?, 0xc002b87f50?}, {0x33d53e0, 0xc0020bceb0}, 0x5c6537c06c010d36, {0xc00291b170, 0x11}, 0x1, 0xc0020bcd20)
	/go/src/github.com/cortexproject/cortex/pkg/querier/worker/scheduler_processor.go:215 +0x924
github.com/cortexproject/cortex/pkg/querier/worker.(*schedulerProcessor).querierLoop.func1()
	/go/src/github.com/cortexproject/cortex/pkg/querier/worker/scheduler_processor.go:162 +0x628
created by github.com/cortexproject/cortex/pkg/querier/worker.(*schedulerProcessor).querierLoop in goroutine 4815
	/go/src/github.com/cortexproject/cortex/pkg/querier/worker/scheduler_processor.go:132 +0x96
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,7 @@
 * [BUGFIX] Query Frontend: Fix samples scanned and peak samples query stats when query hits results cache. #6591
 * [BUGFIX] Query Frontend: Fix panic caused by nil pointer dereference. #6609
 * [BUGFIX] Ingester: Add check to avoid query 5xx when closing tsdb. #6616
+* [BUGFIX] Querier: Fix panic when marshaling QueryResultRequest. #6601
 
 ## 1.19.0 2025-02-27
 

--- a/pkg/querier/stats/stats.go
+++ b/pkg/querier/stats/stats.go
@@ -49,6 +49,16 @@ func IsEnabled(ctx context.Context) bool {
 	return FromContext(ctx) != nil
 }
 
+func (s *QueryStats) Copy() *QueryStats {
+	if s == nil {
+		return nil
+	}
+
+	copied := &QueryStats{}
+	copied.Merge(s)
+	return copied
+}
+
 // AddWallTime adds some time to the counter.
 func (s *QueryStats) AddWallTime(t time.Duration) {
 	if s == nil {

--- a/pkg/querier/stats/stats_test.go
+++ b/pkg/querier/stats/stats_test.go
@@ -8,6 +8,26 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+func TestStats_Copy(t *testing.T) {
+	t.Run("stats is nil", func(t *testing.T) {
+		var stats *QueryStats
+		copied := stats.Copy()
+		assert.Nil(t, copied)
+	})
+	t.Run("stats is not nil", func(t *testing.T) {
+		stats, _ := ContextWithEmptyStats(context.Background())
+		stats.AddWallTime(time.Second)
+		copied := stats.Copy()
+
+		// value should be the same
+		assert.Equal(t, time.Second, copied.LoadWallTime())
+		p1, p2 := &copied, &stats
+
+		// address should be different
+		assert.False(t, p1 == p2)
+	})
+}
+
 func TestStats_WallTime(t *testing.T) {
 	t.Run("add and load wall time", func(t *testing.T) {
 		stats, _ := ContextWithEmptyStats(context.Background())

--- a/pkg/querier/worker/scheduler_processor.go
+++ b/pkg/querier/worker/scheduler_processor.go
@@ -39,6 +39,9 @@ func newSchedulerProcessor(cfg Config, handler RequestHandler, log log.Logger, r
 		querierID:      cfg.QuerierID,
 		grpcConfig:     cfg.GRPCClientConfig,
 		targetHeaders:  cfg.TargetHeaders,
+		schedulerClientFactory: func(conn *grpc.ClientConn) schedulerpb.SchedulerForQuerierClient {
+			return schedulerpb.NewSchedulerForQuerierClient(conn)
+		},
 		frontendClientRequestDuration: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
 			Name:    "cortex_querier_query_frontend_request_duration_seconds",
 			Help:    "Time spend doing requests to frontend.",
@@ -72,12 +75,13 @@ type schedulerProcessor struct {
 	frontendPool                  *client.Pool
 	frontendClientRequestDuration *prometheus.HistogramVec
 
-	targetHeaders []string
+	targetHeaders          []string
+	schedulerClientFactory func(conn *grpc.ClientConn) schedulerpb.SchedulerForQuerierClient
 }
 
 // notifyShutdown implements processor.
 func (sp *schedulerProcessor) notifyShutdown(ctx context.Context, conn *grpc.ClientConn, address string) {
-	client := schedulerpb.NewSchedulerForQuerierClient(conn)
+	client := sp.schedulerClientFactory(conn)
 
 	req := &schedulerpb.NotifyQuerierShutdownRequest{QuerierID: sp.querierID}
 	if _, err := client.NotifyQuerierShutdown(ctx, req); err != nil {
@@ -87,7 +91,7 @@ func (sp *schedulerProcessor) notifyShutdown(ctx context.Context, conn *grpc.Cli
 }
 
 func (sp *schedulerProcessor) processQueriesOnSingleStream(ctx context.Context, conn *grpc.ClientConn, address string) {
-	schedulerClient := schedulerpb.NewSchedulerForQuerierClient(conn)
+	schedulerClient := sp.schedulerClientFactory(conn)
 
 	backoff := backoff.New(ctx, processorBackoffConfig)
 	for backoff.Ongoing() {
@@ -211,11 +215,15 @@ func (sp *schedulerProcessor) runRequest(ctx context.Context, logger log.Logger,
 
 	c, err := sp.frontendPool.GetClientFor(frontendAddress)
 	if err == nil {
+		// To prevent querier panic, the panic could happen when the go-routines not-exited
+		// yet in `fetchSeriesFromStores` are increment query-stats while progressing
+		// (*QueryResultRequest).MarshalToSizedBuffer under the same query-stat objects are used.
+		copiedStats := stats.Copy()
 		// Response is empty and uninteresting.
 		_, err = c.(frontendv2pb.FrontendForQuerierClient).QueryResult(ctx, &frontendv2pb.QueryResultRequest{
 			QueryID:      queryID,
 			HttpResponse: response,
-			Stats:        stats,
+			Stats:        copiedStats,
 		})
 	}
 	if err != nil {

--- a/pkg/querier/worker/scheduler_processor_test.go
+++ b/pkg/querier/worker/scheduler_processor_test.go
@@ -1,0 +1,175 @@
+package worker
+
+import (
+	"context"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/weaveworks/common/httpgrpc"
+	"go.uber.org/atomic"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/cortexproject/cortex/pkg/frontend/v2/frontendv2pb"
+	"github.com/cortexproject/cortex/pkg/querier/stats"
+	"github.com/cortexproject/cortex/pkg/scheduler/schedulerpb"
+)
+
+// mock querier request handler
+type mockRequestHandler struct {
+	mock.Mock
+}
+
+func (m *mockRequestHandler) Handle(ctx context.Context, req *httpgrpc.HTTPRequest) (*httpgrpc.HTTPResponse, error) {
+	args := m.Called(ctx, req)
+	return args.Get(0).(*httpgrpc.HTTPResponse), args.Error(1)
+}
+
+type mockFrontendForQuerierServer struct {
+	mock.Mock
+}
+
+func (m *mockFrontendForQuerierServer) QueryResult(_ context.Context, _ *frontendv2pb.QueryResultRequest) (*frontendv2pb.QueryResultResponse, error) {
+	return &frontendv2pb.QueryResultResponse{}, nil
+}
+
+type mockSchedulerForQuerierClient struct {
+	mock.Mock
+}
+
+func (m *mockSchedulerForQuerierClient) QuerierLoop(ctx context.Context, opts ...grpc.CallOption) (schedulerpb.SchedulerForQuerier_QuerierLoopClient, error) {
+	args := m.Called(ctx, opts)
+	return args.Get(0).(schedulerpb.SchedulerForQuerier_QuerierLoopClient), args.Error(1)
+}
+
+func (m *mockSchedulerForQuerierClient) NotifyQuerierShutdown(ctx context.Context, in *schedulerpb.NotifyQuerierShutdownRequest, opts ...grpc.CallOption) (*schedulerpb.NotifyQuerierShutdownResponse, error) {
+	args := m.Called(ctx, in, opts)
+	return args.Get(0).(*schedulerpb.NotifyQuerierShutdownResponse), args.Error(1)
+}
+
+// mock SchedulerForQuerier_QuerierLoopClient
+type mockQuerierLoopClient struct {
+	ctx context.Context
+	mock.Mock
+}
+
+func (m *mockQuerierLoopClient) Send(msg *schedulerpb.QuerierToScheduler) error {
+	args := m.Called(msg)
+	return args.Error(0)
+}
+
+func (m *mockQuerierLoopClient) Recv() (*schedulerpb.SchedulerToQuerier, error) {
+	args := m.Called()
+
+	if fn, ok := args.Get(0).(func() (*schedulerpb.SchedulerToQuerier, error)); ok {
+		return fn()
+	}
+
+	return args.Get(0).(*schedulerpb.SchedulerToQuerier), args.Error(1)
+}
+
+func (m *mockQuerierLoopClient) Header() (metadata.MD, error) {
+	args := m.Called()
+	return args.Get(0).(metadata.MD), args.Error(1)
+}
+
+func (m *mockQuerierLoopClient) Trailer() metadata.MD {
+	args := m.Called()
+	return args.Get(0).(metadata.MD)
+}
+
+func (m *mockQuerierLoopClient) CloseSend() error {
+	args := m.Called()
+	return args.Error(0)
+}
+
+func (m *mockQuerierLoopClient) Context() context.Context {
+	args := m.Called()
+	return args.Get(0).(context.Context)
+}
+
+func (m *mockQuerierLoopClient) SendMsg(msg interface{}) error {
+	args := m.Called(msg)
+	return args.Error(0)
+}
+
+func (m *mockQuerierLoopClient) RecvMsg(msg interface{}) error {
+	args := m.Called(msg)
+	return args.Error(0)
+}
+
+// To show https://github.com/cortexproject/cortex/issues/6599 issue has been resolved
+func Test_ToShowNotPanic_RelatedIssue6599(t *testing.T) {
+	cfg := Config{}
+	frontendAddress := ":50001"
+	userID := "user-1"
+	recvCount := 20000
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
+	defer cancel()
+
+	recvCall := atomic.Uint32{}
+
+	// mocking query scheduler
+	querierLoopClient := &mockQuerierLoopClient{}
+	querierLoopClient.ctx = ctx
+	querierLoopClient.On("Send", mock.Anything).Return(nil)
+	querierLoopClient.On("Context").Return(querierLoopClient.ctx)
+	querierLoopClient.On("Recv").Return(func() (*schedulerpb.SchedulerToQuerier, error) {
+		recvCall.Add(1)
+		if recvCall.Load() <= uint32(recvCount) {
+			return &schedulerpb.SchedulerToQuerier{
+				QueryID:         1,
+				HttpRequest:     &httpgrpc.HTTPRequest{},
+				FrontendAddress: frontendAddress,
+				UserID:          userID,
+				StatsEnabled:    true,
+			}, nil
+		} else {
+			<-querierLoopClient.ctx.Done()
+			return nil, context.Canceled
+		}
+
+	})
+
+	requestHandler := &mockRequestHandler{}
+	requestHandler.On("Handle", mock.Anything, mock.Anything).Run(func(args mock.Arguments) {
+		stat := stats.FromContext(args.Get(0).(context.Context))
+
+		// imitate add query-stat at fetchSeriesFromStores
+		go stat.AddFetchedChunkBytes(10)
+	}).Return(&httpgrpc.HTTPResponse{}, nil)
+
+	sp, _ := newSchedulerProcessor(cfg, requestHandler, log.NewNopLogger(), nil)
+	schedulerClient := &mockSchedulerForQuerierClient{}
+	schedulerClient.On("QuerierLoop", mock.Anything, mock.Anything).Return(querierLoopClient, nil)
+
+	sp.schedulerClientFactory = func(conn *grpc.ClientConn) schedulerpb.SchedulerForQuerierClient {
+		return schedulerClient
+	}
+
+	// frontendForQuerierServer
+	grpcServer := grpc.NewServer()
+	server := &mockFrontendForQuerierServer{}
+	frontendv2pb.RegisterFrontendForQuerierServer(grpcServer, server)
+
+	lis, err := net.Listen("tcp", frontendAddress)
+	require.NoError(t, err)
+	stopChan := make(chan struct{})
+	go func() {
+		defer close(stopChan)
+		if err := grpcServer.Serve(lis); err != nil {
+			return
+		}
+	}()
+	defer func() {
+		grpcServer.GracefulStop()
+		<-stopChan // Wait util stop complete
+	}()
+
+	sp.processQueriesOnSingleStream(ctx, nil, lis.Addr().String())
+}

